Add experimental support for AudioSession Web API

youennf opened the pull request:
#### a5482e6f1937ce88085b0e181423e77c129b010b
<pre>
Add experimental support for AudioSession Web API
<a href="https://bugs.webkit.org/show_bug.cgi?id=248811">https://bugs.webkit.org/show_bug.cgi?id=248811</a>
rdar://problem/103019883

Reviewed by Eric Carlson.

Add experimental support for a Web API to better control AudioSession states, in particular category and interruptions.
Covered by added tests.

* LayoutTests/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/media/audioSession/audioSessionState-expected.txt: Added.
* LayoutTests/media/audioSession/audioSessionState.html: Added.
* LayoutTests/media/audioSession/audioSessionType-expected.txt: Added.
* LayoutTests/media/audioSession/audioSessionType.html: Added.
* LayoutTests/media/audioSession/getUserMedia-expected.txt: Added.
* LayoutTests/media/audioSession/getUserMedia.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp: Added.
(WebCore::fromDOMAudioSessionType):
(WebCore::DOMAudioSession::create):
(WebCore::DOMAudioSession::DOMAudioSession):
(WebCore::DOMAudioSession::~DOMAudioSession):
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
(WebCore::DOMAudioSession::state const):
(WebCore::DOMAudioSession::stop):
(WebCore::DOMAudioSession::activeDOMObjectName const):
(WebCore::DOMAudioSession::virtualHasPendingActivity const):
(WebCore::DOMAudioSession::beginAudioSessionInterruption):
(WebCore::DOMAudioSession::endAudioSessionInterruption):
(WebCore::DOMAudioSession::activeStateChanged):
(WebCore::DOMAudioSession::scheduleStateChangeEvent):
* Source/WebCore/Modules/audiosession/DOMAudioSession.h: Added.
* Source/WebCore/Modules/audiosession/DOMAudioSession.idl: Added.
* Source/WebCore/Modules/audiosession/Navigator+AudioSession.idl: Added.
* Source/WebCore/Modules/audiosession/NavigatorAudioSession.cpp: Added.
(WebCore::NavigatorAudioSession::NavigatorAudioSession):
(WebCore::NavigatorAudioSession::audioSession):
(WebCore::NavigatorAudioSession::from):
(WebCore::NavigatorAudioSession::supplementName):
* Source/WebCore/Modules/audiosession/NavigatorAudioSession.h: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::getUserMedia):
* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::allow):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/ActiveDOMObject.cpp:
(WebCore::ActiveDOMObject::queueTaskToDispatchEventInternal):
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::tryToSetActive):
(WebCore::AudioSession::activeStateChanged):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::configurationChanged):

Canonical link: <a href="https://commits.webkit.org/257552@main">https://commits.webkit.org/257552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e81166ebb342aca03c220645fc5e2b6fe28f23a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8490 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32409 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108676 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168922 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85808 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91781 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106603 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105048 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90384 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33823 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21730 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89989 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2367 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23247 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85828 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2265 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28873 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5190 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8032 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42725 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88695 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3908 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19841 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/14f8daa8effb9ed00a73f7e250b405998edf82f0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2abc886a943b52d9b7186633d6b6074db20f1ff6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a931566846d8ed3c46fb8aed6e843f3ff0a8a678. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/700ab701dfd3b4346dbd0d7d0368f38b1bae5b91. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/224094700aa6facdd5b0722be7fa094b3a8bfa77. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f7ab4b15e308060e177b3f384d18c8d127db1142. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5bc0ae1be589f0c36f8b77deb6c288d560aceebf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

eric-carlson commented on LayoutTests/media/audioSession/getUserMedia.html:
```diff
@@ -0,0 +1,40 @@
+
+    navigator.audioSession.type = "playback";
+    await navigator.mediaDevices.getUserMedia({audio:true}).then(assert_unreached, () => { });
+    localStream = await navigator.mediaDevices.getUserMedia({video:true});
+    localStream.getTracks().forEach(track => track.stop());
```
Is this to verify that it is possible to create a video-only stream when the session type is `playback`? If so, maybe add a timeout so the test fails rather than just timing out.

In either case, I think it would be helpful to have a comment explaining what is happening.

youennf replied:
Yes, this is to verify camera access is not tied to audioSession type, I'll add a comment.
I do not think the test will time out if it ever happened, probably the promise would reject (or the camera capture would start but fail, in which case we could beef up the test by checking onended.

eric-carlson commented on Source/WebCore/Modules/audiosession/DOMAudioSession.cpp:
```diff
@@ -0,0 +1,138 @@
+void DOMAudioSession::setType(Type type)
+{
+    s_type = type;
+    AudioSession::sharedSession().setCategoryOverride(fromDOMAudioSessionType(type));
```
Should `setCategoryOverride` call `updateSessionState()` when the category is set to `auto` so we set the category to whatever is appropriate for the current state?

eric-carlson reviewed (approved):
r=me

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3bb89e4000aa8d52544571a5050015bf363cf364. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/7190

eric-carlson commented on Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
```diff
@@ -91,6 +91,13 @@ void PlatformMediaSessionManager::updateNowPlayingInfoIfNecessary()
 
+void PlatformMediaSessionManager::updateAudioSessionCategoryIfNecessary()
+{
+    fprintf(stderr, "PlatformMediaSessionManager::updateAudioSessionCategoryIfNecessary %p\n", PlatformMediaSessionManager::sharedManagerIfExists());
```
Oops, don't land with this!

youennf replied:
Gasp, I did not see your comment before merging... will fix it shortly

youennf replied:
https://github.com/WebKit/WebKit/pull/7319

youennf committed:
Add experimental support for AudioSession Web API
https://bugs.webkit.org/show_bug.cgi?id=248811
rdar://problem/103019883

Reviewed by Eric Carlson.

Add experimental support for a Web API to better control AudioSession states, in particular category and interruptions.
Covered by added tests.

* LayoutTests/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/media/audioSession/audioSessionState-expected.txt: Added.
* LayoutTests/media/audioSession/audioSessionState.html: Added.
* LayoutTests/media/audioSession/audioSessionType-expected.txt: Added.
* LayoutTests/media/audioSession/audioSessionType.html: Added.
* LayoutTests/media/audioSession/getUserMedia-expected.txt: Added.
* LayoutTests/media/audioSession/getUserMedia.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/audiosession/DOMAudioSession.cpp: Added.
(WebCore::fromDOMAudioSessionType):
(WebCore::DOMAudioSession::create):
(WebCore::DOMAudioSession::DOMAudioSession):
(WebCore::DOMAudioSession::~DOMAudioSession):
(WebCore::DOMAudioSession::setType):
(WebCore::DOMAudioSession::type const):
(WebCore::DOMAudioSession::state const):
(WebCore::DOMAudioSession::stop):
(WebCore::DOMAudioSession::activeDOMObjectName const):
(WebCore::DOMAudioSession::virtualHasPendingActivity const):
(WebCore::DOMAudioSession::beginAudioSessionInterruption):
(WebCore::DOMAudioSession::endAudioSessionInterruption):
(WebCore::DOMAudioSession::activeStateChanged):
(WebCore::DOMAudioSession::scheduleStateChangeEvent):
* Source/WebCore/Modules/audiosession/DOMAudioSession.h: Added.
* Source/WebCore/Modules/audiosession/DOMAudioSession.idl: Added.
* Source/WebCore/Modules/audiosession/Navigator+AudioSession.idl: Added.
* Source/WebCore/Modules/audiosession/NavigatorAudioSession.cpp: Added.
(WebCore::NavigatorAudioSession::NavigatorAudioSession):
(WebCore::NavigatorAudioSession::audioSession):
(WebCore::NavigatorAudioSession::from):
(WebCore::NavigatorAudioSession::supplementName):
* Source/WebCore/Modules/audiosession/NavigatorAudioSession.h: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::getUserMedia):
* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::allow):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/ActiveDOMObject.cpp:
(WebCore::ActiveDOMObject::queueTaskToDispatchEventInternal):
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/platform/audio/AudioSession.cpp:
(WebCore::AudioSession::tryToSetActive):
(WebCore::AudioSession::activeStateChanged):
* Source/WebCore/platform/audio/AudioSession.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.cpp:
(WebKit::RemoteAudioSession::configurationChanged):

Canonical link: https://commits.webkit.org/257552@main
